### PR TITLE
Fix convolution compiler warnings and char casts

### PR DESCRIPTION
**Pull request by felixbuenemann**
This fixes lower than zero comparison of unsigned int / short in `im_aconv.c` which would generate a compiler warning in clang.

I also found two places in `im_aconv.c` and `im_aconvsep.c` where supposedly a clipping to signed char is done, but the ckip to unsigned macro was used.

Please review to ensure those changes are correct.

### CODE REVIEW

File: libvips/convolution/im_aconv.c [hunk 1]
```diff
@@ -1005,9 +1005,7 @@ aconv_horizontal( Boxes *boxes, IMAGE *in, IMAGE *out )
 
 #define CLIP_UCHAR( V ) \
 G_STMT_START { \
-	if( (V) < 0 ) \
-		(V) = 0; \
-	else if( (V) > UCHAR_MAX ) \
+	if( (V) > UCHAR_MAX ) \
```
Comment (jcupitt) [inline libvips/convolution/im_aconv.c]: I don't think this is right, it needs to clip <0 ints to 0 when casting down to uchar.

Perhaps your warning is triggered by L1146? That's using uchar for a char case, I'm not sure why. Perhaps changing that line instead would be safer?

Comment (jcupitt) [inline libvips/convolution/im_aconv.c]: Ooop, I see you caught that one too. Is this change still necessary after your second patch?

**Comment by jcupitt** [comment]
`aconv` is unfinished :-( There are some obvious cases it fails to handle, and it needs to be vips8-ed.

I wrote the first 90% of it in a state of great excitement, thinking I'd come up with a new way of doing very fast approximate convolution. When I got it basically working I did the literature search and found a paper from 1986 describing (almost) exactly the same thing. That took the wind from my sails and I never did the final 10%, unfortunately. 

Plain `conv` needs vips8-ing too. And the way it handles accumulation of intermediate values in Orc mode can be improved quite a bit, `reducev` is much better.

File: libvips/convolution/im_aconv.c (hunk 1 above)
Comment (felixbuenemann) [inline libvips/convolution/im_aconv.c]: Well checking an unsigned int for < 0 is a nop. If you subtract 1 from a uint32 you get 2^32-1 so it'll never be smaller than 0. If you wanted to use that code as is you'd have to use signed ints for yor sum arg to VCONV.

**Comment by felixbuenemann** [comment]
I'm not using the convolution code, I just noticed that it generated lots of compiler warnings and looked into them. While reading through the code I also noticed the char vs uchar clipping macro problem.

File: libvips/convolution/im_aconv.c (hunk 1 above)
Comment (felixbuenemann) [inline libvips/convolution/im_aconv.c]: The reason the warnings are generated, is because the clipping macros are used on unsigned types, were the < 0 check does nothing. If those macros were used on signed inputs the check would be correct.

**Comment by jcupitt** [comment]
Hi again Felix, what flags are you using for clang? I tried with:

```
CC=clang CFLAGS="-O3 -Wall" CXX=clang++ CXXFLAGS="-Wall -O3" ./configure --prefix=/home/john/vips --without-magick
```

and I see a clean build. Is there a better enable-warnings flag?

**Comment by felixbuenemann** [comment]
I didn't give any custom flags. I'm probably using a different version of clang here on OS X 10.11 then what's on your system. I can get you the exact warning.

**Comment by felixbuenemann** [comment]
Here's the command and generated warnings (see the underlinings when scrolled way to the right):

```
clang -v
Apple LLVM version 7.3.0 (clang-703.0.31)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

libtool: compile:  /usr/local/opt/ccache/libexec/clang -DHAVE_CONFIG_H -I. -I../.. -I../../libvips/include -DG_DISABLE_ASSERT -DG_DISABLE_CHECKS -I/usr/local/opt/gettext/include -I/usr/local/Cellar/webp/HEAD/include -I/usr/local/Cellar/poppler/0.44.0/include/poppler/glib -I/usr/local/Cellar/poppler/0.44.0/include/poppler -I/usr/local/Cellar/pixman/0.34.0/include/pixman-1 -I/usr/local/Cellar/pango/1.38.1/include/pango-1.0 -I/usr/local/Cellar/orc/0.4.25/include/orc-0.4 -I/usr/local/Cellar/openslide/3.4.1_1/include/openslide -I/usr/local/Cellar/openexr/2.2.0/include/OpenEXR -I/usr/local/Cellar/little-cms2/2.7/include -I/usr/local/Cellar/libtiff/4.0.6_1/include -I/usr/local/Cellar/librsvg/2.40.15/include/librsvg-2.0 -I/usr/local/Cellar/libpng/1.6.21/include/libpng16 -I/usr/local/Cellar/libmatio/1.5.2_1/include -I/usr/local/Cellar/libgsf/DEV/include/libgsf-1 -I/usr/local/Cellar/libexif/0.6.21/include -I/usr/local/Cellar/imagemagick/6.9.4-3/include/ImageMagick-6 -I/usr/local/Cellar/ilmbase/2.2.0/include/OpenEXR -I/usr/local/Cellar/harfbuzz/1.2.6/include/harfbuzz -I/usr/local/Cellar/glib/2.46.2/lib/glib-2.0/include -I/usr/local/Cellar/glib/2.46.2/include/glib-2.0 -I/usr/local/Cellar/gdk-pixbuf/2.32.3/include/gdk-pixbuf-2.0 -I/usr/local/Cellar/freetype/2.6.3/include/freetype2 -I/usr/local/Cellar/fontconfig/2.11.1_2/include -I/usr/local/Cellar/fftw/3.3.4_1/include -I/usr/local/Cellar/cfitsio/3.370/include -I/usr/local/Cellar/cairo/1.14.6_1/include/cairo -I/usr/include/libxml2 -D_THREAD_SAFE -D_REENTRANT -DMAGICKCORE_QUANTUM_DEPTH=16 -DMAGICKCORE_HDRI_ENABLE=0 -I/usr/local/opt/mozjpeg/include -g -O2 -MT im_aconv.lo -MD -MP -MF .deps/im_aconv.Tpo -c im_aconv.c  -fno-common -DPIC -o .libs/im_aconv.o
im_aconv.c:1136:817: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
   do { for( x = 0; x < sz; x++ ) { unsigned int *seq_sum = (unsigned int *) seq->sum; unsigned int *p; unsigned char *q; unsigned int sum; p = x * boxes->n_hline + (unsigned int *) ((ir)->data + ((r->top)-(ir)->valid.top) * ((size_t)((ir)->bpl)) + ((r->left)-(ir)->valid.left) * ((vips_format_sizeof(((ir)->im)->BandFmt)) * ((ir)->im)->Bands)); q = x + (unsigned char *) ((or)->data + ((r->top)-(or)->valid.top) * ((size_t)((or)->bpl)) + ((r->left)-(or)->valid.left) * ((vips_format_sizeof(((or)->im)->BandFmt)) * ((or)->im)->Bands)); sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] = 0; for( k = boxes->vline[z].start; k < boxes->vline[z].end; k++ ) seq_sum[z] += p[k * istride + boxes->vline[z].band]; sum += boxes->vline[z].factor * seq_sum[z]; } sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (127*2 +1) ) (sum) = (127*2 +1); } while (0); *q = sum; q += ostride; for( y = 1; y < r->height; y++ ) { sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] += p[seq->end[z]]; seq_sum[z] -= p[seq->start[z]]; sum += boxes->vline[z].factor * seq_sum[z]; } p += istride; sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (127*2 +1) ) (sum) = (127*2 +1); } while (0); *q = sum; q += ostride; } } } while (0);
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          ~~~~~ ^ ~
im_aconv.c:1136:1180: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
   do { for( x = 0; x < sz; x++ ) { unsigned int *seq_sum = (unsigned int *) seq->sum; unsigned int *p; unsigned char *q; unsigned int sum; p = x * boxes->n_hline + (unsigned int *) ((ir)->data + ((r->top)-(ir)->valid.top) * ((size_t)((ir)->bpl)) + ((r->left)-(ir)->valid.left) * ((vips_format_sizeof(((ir)->im)->BandFmt)) * ((ir)->im)->Bands)); q = x + (unsigned char *) ((or)->data + ((r->top)-(or)->valid.top) * ((size_t)((or)->bpl)) + ((r->left)-(or)->valid.left) * ((vips_format_sizeof(((or)->im)->BandFmt)) * ((or)->im)->Bands)); sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] = 0; for( k = boxes->vline[z].start; k < boxes->vline[z].end; k++ ) seq_sum[z] += p[k * istride + boxes->vline[z].band]; sum += boxes->vline[z].factor * seq_sum[z]; } sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (127*2 +1) ) (sum) = (127*2 +1); } while (0); *q = sum; q += ostride; for( y = 1; y < r->height; y++ ) { sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] += p[seq->end[z]]; seq_sum[z] -= p[seq->start[z]]; sum += boxes->vline[z].factor * seq_sum[z]; } p += istride; sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (127*2 +1) ) (sum) = (127*2 +1); } while (0); *q = sum; q += ostride; } } } while (0);
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ~~~~~ ^ ~
im_aconv.c:1139:821: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
   do { for( x = 0; x < sz; x++ ) { unsigned int *seq_sum = (unsigned int *) seq->sum; unsigned short *p; unsigned char *q; unsigned int sum; p = x * boxes->n_hline + (unsigned short *) ((ir)->data + ((r->top)-(ir)->valid.top) * ((size_t)((ir)->bpl)) + ((r->left)-(ir)->valid.left) * ((vips_format_sizeof(((ir)->im)->BandFmt)) * ((ir)->im)->Bands)); q = x + (unsigned char *) ((or)->data + ((r->top)-(or)->valid.top) * ((size_t)((or)->bpl)) + ((r->left)-(or)->valid.left) * ((vips_format_sizeof(((or)->im)->BandFmt)) * ((or)->im)->Bands)); sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] = 0; for( k = boxes->vline[z].start; k < boxes->vline[z].end; k++ ) seq_sum[z] += p[k * istride + boxes->vline[z].band]; sum += boxes->vline[z].factor * seq_sum[z]; } sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (127*2 +1) ) (sum) = (127*2 +1); } while (0); *q = sum; q += ostride; for( y = 1; y < r->height; y++ ) { sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] += p[seq->end[z]]; seq_sum[z] -= p[seq->start[z]]; sum += boxes->vline[z].factor * seq_sum[z]; } p += istride; sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (127*2 +1) ) (sum) = (127*2 +1); } while (0); *q = sum; q += ostride; } } } while (0);
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              ~~~~~ ^ ~
im_aconv.c:1139:1184: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
   do { for( x = 0; x < sz; x++ ) { unsigned int *seq_sum = (unsigned int *) seq->sum; unsigned short *p; unsigned char *q; unsigned int sum; p = x * boxes->n_hline + (unsigned short *) ((ir)->data + ((r->top)-(ir)->valid.top) * ((size_t)((ir)->bpl)) + ((r->left)-(ir)->valid.left) * ((vips_format_sizeof(((ir)->im)->BandFmt)) * ((ir)->im)->Bands)); q = x + (unsigned char *) ((or)->data + ((r->top)-(or)->valid.top) * ((size_t)((or)->bpl)) + ((r->left)-(or)->valid.left) * ((vips_format_sizeof(((or)->im)->BandFmt)) * ((or)->im)->Bands)); sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] = 0; for( k = boxes->vline[z].start; k < boxes->vline[z].end; k++ ) seq_sum[z] += p[k * istride + boxes->vline[z].band]; sum += boxes->vline[z].factor * seq_sum[z]; } sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (127*2 +1) ) (sum) = (127*2 +1); } while (0); *q = sum; q += ostride; for( y = 1; y < r->height; y++ ) { sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] += p[seq->end[z]]; seq_sum[z] -= p[seq->start[z]]; sum += boxes->vline[z].factor * seq_sum[z]; } p += istride; sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (127*2 +1) ) (sum) = (127*2 +1); } while (0); *q = sum; q += ostride; } } } while (0);
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ~~~~~ ^ ~
im_aconv.c:1153:818: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
  do { for( x = 0; x < sz; x++ ) { unsigned int *seq_sum = (unsigned int *) seq->sum; unsigned int *p; unsigned short *q; unsigned int sum; p = x * boxes->n_hline + (unsigned int *) ((ir)->data + ((r->top)-(ir)->valid.top) * ((size_t)((ir)->bpl)) + ((r->left)-(ir)->valid.left) * ((vips_format_sizeof(((ir)->im)->BandFmt)) * ((ir)->im)->Bands)); q = x + (unsigned short *) ((or)->data + ((r->top)-(or)->valid.top) * ((size_t)((or)->bpl)) + ((r->left)-(or)->valid.left) * ((vips_format_sizeof(((or)->im)->BandFmt)) * ((or)->im)->Bands)); sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] = 0; for( k = boxes->vline[z].start; k < boxes->vline[z].end; k++ ) seq_sum[z] += p[k * istride + boxes->vline[z].band]; sum += boxes->vline[z].factor * seq_sum[z]; } sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (32767 *2 +1) ) (sum) = (32767 *2 +1); } while (0); *q = sum; q += ostride; for( y = 1; y < r->height; y++ ) { sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] += p[seq->end[z]]; seq_sum[z] -= p[seq->start[z]]; sum += boxes->vline[z].factor * seq_sum[z]; } p += istride; sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (32767 *2 +1) ) (sum) = (32767 *2 +1); } while (0); *q = sum; q += ostride; } } } while (0);
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           ~~~~~ ^ ~
im_aconv.c:1153:1187: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
  do { for( x = 0; x < sz; x++ ) { unsigned int *seq_sum = (unsigned int *) seq->sum; unsigned int *p; unsigned short *q; unsigned int sum; p = x * boxes->n_hline + (unsigned int *) ((ir)->data + ((r->top)-(ir)->valid.top) * ((size_t)((ir)->bpl)) + ((r->left)-(ir)->valid.left) * ((vips_format_sizeof(((ir)->im)->BandFmt)) * ((ir)->im)->Bands)); q = x + (unsigned short *) ((or)->data + ((r->top)-(or)->valid.top) * ((size_t)((or)->bpl)) + ((r->left)-(or)->valid.left) * ((vips_format_sizeof(((or)->im)->BandFmt)) * ((or)->im)->Bands)); sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] = 0; for( k = boxes->vline[z].start; k < boxes->vline[z].end; k++ ) seq_sum[z] += p[k * istride + boxes->vline[z].band]; sum += boxes->vline[z].factor * seq_sum[z]; } sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (32767 *2 +1) ) (sum) = (32767 *2 +1); } while (0); *q = sum; q += ostride; for( y = 1; y < r->height; y++ ) { sum = 0; for( z = 0; z < n_vline; z++ ) { seq_sum[z] += p[seq->end[z]]; seq_sum[z] -= p[seq->start[z]]; sum += boxes->vline[z].factor * seq_sum[z]; } p += istride; sum = (sum + boxes->rounding) / boxes->area; do { if( (sum) < 0 ) (sum) = 0; else if( (sum) > (32767 *2 +1) ) (sum) = (32767 *2 +1); } while (0); *q = sum; q += ostride; } } } while (0);
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ~~~~~ ^ ~
6 warnings generated.
```

**Comment by felixbuenemann** [comment]
Btw. this also comes up in the resample `mapim.c`:

```
mapim.c:276:146: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
   { unsigned int * restrict p1 = (unsigned int *) p; for( x = 0; x < r->width; x++ ) { unsigned int px = p1[0]; unsigned int py = p1[1]; if( px < 0 || px >= resample->in->Xsize || py < 0 || py >= resample->in->Ysize ) { for( z = 0; z < ps; z++ ) q[z] = 0; } else interpolate( mapim->interpolate, q, ir[0], px + window_offset, py + window_offset ); p1 += 2; q += ps; } }; break;
                                                                                                                                              ~~ ^ ~
mapim.c:276:185: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
   { unsigned int * restrict p1 = (unsigned int *) p; for( x = 0; x < r->width; x++ ) { unsigned int px = p1[0]; unsigned int py = p1[1]; if( px < 0 || px >= resample->in->Xsize || py < 0 || py >= resample->in->Ysize ) { for( z = 0; z < ps; z++ ) q[z] = 0; } else interpolate( mapim->interpolate, q, ir[0], px + window_offset, py + window_offset ); p1 += 2; q += ps; } }; break;
                                                                                                                                                                                     ~~ ^ ~
2 warnings generated.
````

**Comment by jcupitt** [comment]
Huh, so strange, I still don't see any warnings here, even with `-Wtautological-compare`. I'll try on my Mac. I guess it's a clang version thing.

```
$ clang --version
clang version 3.8.0-2ubuntu3 (tags/RELEASE_380/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by felixbuenemann** [comment]
@jcupitt If you don't like the warning fixes, you could just cherry-pick the `Fix wrong char clipping in convolution op` commit.

**Comment by jcupitt** [comment]
This is so strange, I get no warnings on OS X 10.11.4 either, I tried a range of configure flags and options, all build aconv cleanly. Even explicitly specifying `-Wtautological-compare` and not using the gcc wrapper. I do see some annoying gettext warnings, and some warning from swig code, but that's all. 

```
john@katamata ~/GIT/libvips (master) $ clang --version
Apple LLVM version 7.3.0 (clang-703.0.31)
```

I suppose I'm nervous making changes to a complicated bit of unfinished code with no tests. Your changes look correct, but I can't prove or test that they are correct. Maybe aconv should be moved to `deprecated/` until it is finished.

**Comment by felixbuenemann** [comment]
Strange, I'm using the same version:

```
$ clang -v
Apple LLVM version 7.3.0 (clang-703.0.31)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

But as I said, the unsigned < 0 checks are no-ops, but the wrong cast macro seems like a bug, so you might wanna ignore the warning fixes and only cherry-pick the bugfix commit.

**Comment by jcupitt** [comment]
I think the right thing to do is rewrite the convolution operators as vips8 classes.

I've started a branch here:

https://github.com/jcupitt/libvips/tree/rewrite-convf

I've done the float one (easy), I'm doing the int one now (harder, but it should be possible to speed up the vector path a bit), then I'll attack the approximate one and try to finish it.

**Comment by jcupitt** [comment]
Can I close this, Felix? I've made an issue for the im_aconv() rewrite instead.

**Comment by felixbuenemann** [comment]
Sure, closed.
